### PR TITLE
Feature/groups api

### PR DESCRIPTION
**Pull request by esotrope**

### CODE REVIEW

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package org.opentestsystem.rdw.reporting.assessment;
+
+public final class Subjects {
+
+    private Subjects() {
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: nit: i tend to put the private ctor for utility classes at the bottom; the reason is that i want the user to see the useful stuff first (at the top). Also, i don't think utility classes are quite obvious enough without javadocs; i don't think documenting the ctor is necessary but it doesn't hurt to throw on a `hide utility class constructor`.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.opentestsystem.rdw.reporting.assessment;
+
+public final class Subjects {
```
Comment (agorina) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: Just curious: why does it need to be ```final```?

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    private Subjects() {
+    }
+
+    /**
+     * This method will normalize subject IDs into a standard non null format
+     *
+     * @param entityId the subject ID to normalize
+     * @return the value of subject ID or <code>0L</code> if the given subject ID is null
+     */
+    public static long entityIdToCode(final Long entityId) {
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: A convention i've seen is to use `safe` in the method to indicate it is eating nulls, e.g. `safeEntityId(entityId)`. Another alternative is `orValue`, so in this case `entityIdOrZero(entityId)` but that is kind of hokey. Not gonna lose sleep if you ignore me, just thinking "out loud".

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java (hunk 2 above)
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: This is just a part of the utility class pattern - its to show that the class isn't intended for instance-level usage and extension/inheritance.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java (hunk 3 above)
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: I like safeEntityId

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupRepository.java [hunk 4]
```diff
@@ -1,39 +1,37 @@
...
 import org.springframework.stereotype.Repository;
 
+import javax.validation.constraints.NotNull;
 import java.util.List;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 @Repository
-public class GroupRepositoryImpl implements GroupRepository {
+public class DefaultGroupRepository implements GroupRepository {
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupRepository.java]: JdbcGroupRepository?  Again, just thinking out loud here.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+public class DefaultGroupService implements GroupService {
+
+    private final GroupRepository groupRepository;
+
+    @Autowired
+    public DefaultGroupService(final @NotNull GroupRepository groupRepository) {
+        this.groupRepository = groupRepository;
+    }
+
+    public Set<Group> getGroups(final User user) {
```
Comment (agorina) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java]: Nit: probably ```@NotNull``` would be good here.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupRepository.java (hunk 4 above)
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupRepository.java]: i can roll with that

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java [hunk 6]
```diff
@@ -1,11 +1,20 @@
...
+
+import com.google.common.base.MoreObjects;
 
 public class Group {
+
     private long id;
     private String name;
     private String schoolName;
-    private String subjectId; //TODO:?
+    private long subject;
```
Comment (agorina) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java]: This is really subjectId. Do you view it differently?

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.validation.constraints.NotNull;
+import java.util.Set;
+
+import static com.google.common.collect.Sets.newHashSet;
+
+@Service
+public class DefaultGroupService implements GroupService {
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java]: RepositoryBackedGroupService?  A convention developed at a company i used to work for.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java (hunk 6 above)
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java]: Do we want to leave this subjectId? Just for clarity?

Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java]: yeah it is - but as far as the client is concerned this is just what a subject is - a numeric value. the client will never have to distinguish subjectCode from subjectId etc its just a numeric value to associate with a subject. I'm ok with either way though. Might be appropriate to make it subjectId because it kind of self-documents where the value comes from

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java [hunk 8]
```diff
@@ -24,8 +21,8 @@ public GroupController(final GroupService service) {
     }
 
     @GetMapping
-    @ResponseStatus(HttpStatus.OK)
-    public List<Group> getGroups(@AuthenticationPrincipal final User user) {
-        return service.findAll(user);
+    public Set<Group> getGroups(@AuthenticationPrincipal final User user) {
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java]: No objection, just curious why? Sometimes Set's are more problematic than List's but if the order is meaningless/nonpredictable this makes sense.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java (hunk 7 above)
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java]: RepositoryBacked, I feel like there should be a better way to say the same thing but maybe not

File: webapp/src/main/resources/application.sql.yml [hunk 9]
```diff
@@ -7,6 +7,17 @@ sql:
             join school s on s.id = sg.school_id
             where user_login=:user_login
 
+    findAllForUsername: >-
+      select
+      	  sg.id,
```
Comment (malaffoon) [inline webapp/src/main/resources/application.sql.yml]: nit: consistent indenting

File: webapp/src/main/resources/application.sql.yml [hunk 10]
```diff
@@ -7,6 +7,17 @@ sql:
             join school s on s.id = sg.school_id
             where user_login=:user_login
```
Comment (agorina) [inline webapp/src/main/resources/application.sql.yml]: The PR does not show the full query above, but I think it is pretty much the same. Should it be removed?

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java (hunk 8 above)
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java]: this is kind of an intent thing - conceptually you are looking up a set of groups is why. It also informs the client that the values are not sorted yet and the client needs to sort them. If we return a list it may be assumed that it is pre-sorted in some meaningful way

File: webapp/src/main/resources/application.sql.yml (hunk 10 above)
Comment (esotrope) [inline webapp/src/main/resources/application.sql.yml]: yeah I can clean it up

File: webapp/src/main/resources/application.sql.yml (hunk 9 above)
Comment (esotrope) [inline webapp/src/main/resources/application.sql.yml]: is that two indents for continuation?

File: webapp/src/test/java/org/opentestsystem/rdw/reporting/group/GroupRepositoryIT.java [hunk 11]
```diff
@@ -32,7 +33,12 @@
...
-        assertThat(repository.findAll("user-login").size()).isEqualTo(2);
-        //TODO: add more verifications
+        assertThat(repository.findAllForUsername("user-login"))
+                .usingFieldByFieldElementComparator()
+                .containsAll(newArrayList(
+                        Group.builder().id(-2).name("IT Student Group 2").schoolName("Test School 1").subject(-1L).build(),
+                        Group.builder().id(-1).name("IT Student Group 1").schoolName("Test School 1").subject(-1L).build()
+                ));
     }
+
```
Comment (agorina) [inline webapp/src/test/java/org/opentestsystem/rdw/reporting/group/GroupRepositoryIT.java]: Suggest to add a test with NULL subject, just because there is an extra step in the Repository code that converts it.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java (hunk 2 above)
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: +1; it seems to be a best practice to make utility classes `final` with a private ctor instead of `abstract`. Some folk consider `abstract` to mean, "this class is intended to be extended" so they get confused by abstract utility classes. And, if it isn't abstract, the ctor has to be private.

File: webapp/src/test/java/org/opentestsystem/rdw/reporting/group/GroupRepositoryIT.java (hunk 11 above)
Comment (esotrope) [inline webapp/src/test/java/org/opentestsystem/rdw/reporting/group/GroupRepositoryIT.java]: good point

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java (hunk 7 above)
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/group/DefaultGroupService.java]: Repositorified?  ;)

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java (hunk 2 above)
Comment (agorina) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/assessment/Subjects.java]: Thank you for the explanation.

File: webapp/src/main/resources/application.sql.yml (hunk 10 above)
Comment (agorina) [inline webapp/src/main/resources/application.sql.yml]: Thank you.

File: webapp/src/main/resources/application.sql.yml (hunk 9 above)
Comment (malaffoon) [inline webapp/src/main/resources/application.sql.yml]: Oh i see, the sg.name is a continuation of the fields. Hmm, personally i'd avoid that issue and do something like:
```sql
      select sg.id, sg.name, school.name as school_name, subject_id
        from user_student_group usg
          join ...
          join ...
        where user_login=:user_login
```

Comment (esotrope) [inline webapp/src/main/resources/application.sql.yml]: 👍
